src/layout/node.ts

```diff
@@ -17,29 +17,44 @@ export default function(node: Node, source: Source): void {
     }
 
     let call = dom.has(node) ? "update" : "add";
+    let parent = node.parentElement ? node.parentElement : null;
     switch (node.nodeType) {
         case Node.DOCUMENT_TYPE_NODE:
             let doctype = node as DocumentType;
             let docAttributes = { name: doctype.name, publicId: doctype.publicId, systemId: doctype.systemId };
             let docData = { tag: "*D", attributes: docAttributes };
-            dom[call](node, docData, source);
+            dom[call](node, parent, docData, source);
+            break;
+        case Node.DOCUMENT_FRAGMENT_NODE:
+            let shadowRoot = (node as ShadowRoot);
+            if (shadowRoot.host) {
+                // See: https://wicg.github.io/construct-stylesheets/
+                // For more details on adoptedStyleSheets for shadow dom
+                let style = "";
+                let adoptedStyleSheets: CSSStyleSheet[] = "adoptedStyleSheets" in shadowRoot ? shadowRoot["adoptedStyleSheets"] : [];
+                for (let styleSheet of adoptedStyleSheets) { style += getCssRules(styleSheet); }
+                let fragementData = { tag: "*S", attributes: { style } };
```

Comment:
>"*S" [](start = 43, length = 4)

See comment elsewhere but just give these constant string names.

Reply:
They are sent over the wire and follow the same pattern as previous tags e.g. *D, *B. I'll make a const enum so it's easier to follow them while saving bytes over the wire.

Reply:
Changed into const enums